ddl/partition.go

```diff
@@ -2154,6 +2159,363 @@ func (w *worker) onExchangeTablePartition(d *ddlCtx, t *meta.Meta, job *model.Jo
+		backfillWorker: newBackfillWorker(jc.ddlJobCtx, sessCtx, id, t, reorgInfo, typeReorgPartitionWorker),
+		metricCounter:  metrics.BackfillTotalCounter.WithLabelValues(metrics.GenerateReorgLabel("reorg_partition_rate", reorgInfo.SchemaName, t.Meta().Name.String())),
+		rowDecoder:     decoder.NewRowDecoder(t, t.WritableCols(), decodeColMap),
+		rowMap:         make(map[int64]types.Datum, len(decodeColMap)),
+		jobContext:     jc,
+	}
+}
+
+func (w *reorgPartitionWorker) BackfillDataInTxn(handleRange reorgBackfillTask) (taskCtx backfillTaskContext, errInTxn error) {
+	oprStartTime := time.Now()
+	ctx := kv.WithInternalSourceType(context.Background(), w.jobContext.ddlJobSourceType())
+	errInTxn = kv.RunInNewTxn(ctx, w.sessCtx.GetStore(), true, func(ctx context.Context, txn kv.Transaction) error {
+		taskCtx.addedCount = 0
+		taskCtx.scanCount = 0
+		txn.SetOption(kv.Priority, w.priority)
+		if tagger := w.reorgInfo.d.getResourceGroupTaggerForTopSQL(w.reorgInfo.Job); tagger != nil {
+			txn.SetOption(kv.ResourceGroupTagger, tagger)
+		}
+
+		rowRecords, nextKey, taskDone, err := w.fetchRowColVals(txn, handleRange)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		taskCtx.nextKey = nextKey
+		taskCtx.done = taskDone
+
+		warningsMap := make(map[errors.ErrorID]*terror.Error, len(rowRecords))
+		warningsCountMap := make(map[errors.ErrorID]int64, len(rowRecords))
+		for _, prr := range rowRecords {
+			taskCtx.scanCount++
+
+			err = txn.Set(prr.key, prr.vals)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			taskCtx.addedCount++
+			if prr.warning != nil {
+				if _, ok := warningsCountMap[prr.warning.ID()]; ok {
+					warningsCountMap[prr.warning.ID()]++
+				} else {
+					warningsCountMap[prr.warning.ID()] = 1
+					warningsMap[prr.warning.ID()] = prr.warning
+				}
+			}
+			// TODO: Future optimization: also write the indexes here?
+			// What if the transaction limit is just enough for a single row, without index?
+			// Hmm, how could that be in the first place?
+			// For now, implement the batch-txn w.addTableIndex,
+			// since it already exists and is in use
+		}
+
+		// Collect the warnings.
+		taskCtx.warnings, taskCtx.warningsCount = warningsMap, warningsCountMap
+
+		// also add the index entries here? And make sure they are not added somewhere else
+
+		return nil
+	})
+	logSlowOperations(time.Since(oprStartTime), "BackfillDataInTxn", 3000)
+
+	return
+}
+
+// Duplicate of updateColumnWorker fetchRowColVals TODO: combine!
+func (w *reorgPartitionWorker) fetchRowColVals(txn kv.Transaction, taskRange reorgBackfillTask) ([]*rowRecord, kv.Key, bool, error) {
+	w.rowRecords = w.rowRecords[:0]
+	startTime := time.Now()
+
+	// taskDone means that the added handle is out of taskRange.endHandle.
+	taskDone := false
+	sysTZ := w.sessCtx.GetSessionVars().StmtCtx.TimeZone
+	reorgedTbl, err := tables.GetReorganizedPartitionedTable(w.table)
+	if err != nil {
+		return nil, nil, true, errors.Trace(err)
+	}
```

Comment:
Extract this out of `fetchRowColVals` so that we don't have to construct the reorganized table for every batch.

Reply:
Fixed, moved to initialisation of `reorgPartitionWorker` / `newReorgPartitionWorker`.